A2rchi/interfaces/chat_app/app.py

```diff
@@ -105,63 +105,79 @@ def __call__(self, history: Optional[List[Tuple[str, str]]], discussion_id: Opti
+            self.lock.acquire()
+            print("INFO - acquired lock file update vectorstore")
+
             self.data_manager.update_vectorstore()
 
-            # convert the history to native A2rchi form (because javascript does not have tuples)
-            history = self.convert_to_chain_history(history)
-
-            # get discussion ID so that the conversation can be saved (It seems that random is no good... TODO)
-            discussion_id = discussion_id or np.random.randint(100000, 999999)
-
-            # run chain to get result
-            if self.number_of_queries < QUERY_LIMIT:
-                result = self.chain(history)
-            else: 
-                # the case where we have exceeded the QUERY LIMIT (built so that we do not overuse the chain)
-                output = "Sorry, our service is currently down due to exceptional demand. Please come again later."
-                return output, discussion_id
-            self.number_of_queries += 1
-            print(f"number of queries is: {self.number_of_queries}")
-
-            # get similarity score to see how close the input is to the source
-            # - low score means very close (it's a distance between embedding vectors approximated
-            #   by an approximate k-nearest neighbors algorithm called HNSW)
-            inp = history[-1][1]
-            score = self.chain.similarity_search(inp)
-
-            # load the present list of sources
-            try:
-                with open(os.path.join(self.data_path, 'sources.yml'), 'r') as file:
-                    sources = yaml.load(file, Loader=yaml.FullLoader)
-            except FileNotFoundError:
-                sources = dict()
-
-            # get the closest source to the document
-            source = None
-            if len(result['source_documents']) > 0:
-                source_hash = result['source_documents'][0].metadata['source']
-                if '/' in source_hash and '.' in source_hash:
-                    source = source_hash.split('/')[-1].split('.')[0]
-
-            # if the score is low enough, include the source as a link, otherwise give just the answer
-            embedding_name = self.config["utils"]["embeddings"]["EMBEDDING_NAME"]
-            similarity_score_reference = self.config["utils"]["embeddings"]["EMBEDDING_CLASS_MAP"][embedding_name]["similarity_score_reference"]
-            if score < similarity_score_reference and source in sources.keys(): 
-                output = "<p>" + result["answer"] + "</p>" + "\n\n<br /><br /><p><a href= " + sources[source] + ">Click here to read more</a></p>"
-            else:
-                output = "<p>" + result["answer"] + "</p>"
+            self.lock.release()
+            print("INFO - released lock file update vectorstore")
+        except Exception as e:
+            print(f"ERROR - {str(e)}")
+        finally:
+            self.lock.release()
```

Comment:
Are we sure about the double release? I remember we discussed this already, but from what I saw it should be avoided to release locks already released. e.g. see in:   https://www.bogotobogo.com/python/Multithread/python_multithreading_Synchronization_Lock_Objects_Acquire_Release.php (can Ctrl+F "The release() method should only be called in the locked state")

Reply:
ibid in update_discussion()

Reply:
Just updated a proposed fix